utils/redis.ts

```diff
@@ -1,69 +1,81 @@
 import { identity, pipe } from "fp-ts/lib/function";
 import * as O from "fp-ts/lib/Option";
 import * as redis from "redis";
-import RedisClustr = require("redis-clustr");
 import { getConfigOrThrow } from "./config";
+
 const config = getConfigOrThrow();
+const DEFAULT_REDIS_PORT = "6379";
+
+export type RedisClient = redis.RedisClientType | redis.RedisClusterType;
 
-const createSimpleRedisClient = (
+export const createSimpleRedisClient = async (
   redisUrl: string,
   password?: string,
   port?: string,
   useTls: boolean = true
-): redis.RedisClient => {
-  const DEFAULT_REDIS_PORT = "6379";
-
+): Promise<RedisClient> => {
   const redisPort: number = parseInt(port || DEFAULT_REDIS_PORT, 10);
-  return redis.createClient({
-    auth_pass: password,
-    host: redisUrl,
-    port: redisPort,
-    tls: useTls ? { servername: redisUrl } : undefined
+  const redisClient = redis.createClient<
+    redis.RedisDefaultModules,
+    Record<string, never>,
+    Record<string, never>
+  >({
+    password,
+    socket: {
+      port: redisPort,
+      tls: useTls
+    },
+    url: redisUrl
   });
+  await redisClient.connect();
+  return redisClient;
 };
 
-const createClusterRedisClient = (
+export const createClusterRedisClient = async (
   redisUrl: string,
   password?: string,
   port?: string
-): redis.RedisClient => {
-  const DEFAULT_REDIS_PORT = "6379";
-
+): Promise<RedisClient> => {
   const redisPort: number = parseInt(port || DEFAULT_REDIS_PORT, 10);
-  return new RedisClustr({
-    redisOptions: {
-      auth_pass: password,
-      tls: {
-        servername: redisUrl
-      }
+  const redisClient = redis.createCluster<
+    redis.RedisDefaultModules,
+    Record<string, never>,
+    Record<string, never>
+  >({
+    defaults: {
+      legacyMode: true,
+      password
     },
-    servers: [
+    rootNodes: [
       {
-        host: redisUrl,
-        port: redisPort
+        url: `${redisUrl}:${redisPort}`
       }
-    ]
-  }) as redis.RedisClient; // Casting RedisClustr with missing typings to RedisClient (same usage).
+    ],
+    useReplicas: true
+  });
+  await redisClient.connect();
+  return redisClient;
 };
 
-export const REDIS_CLIENT = pipe(
-  config.isProduction,
-  O.fromPredicate<boolean>(identity),
-  O.chainNullableK(_ => config.REDIS_CLUSTER_ENABLED),
-  O.chain(O.fromPredicate(identity)),
-  O.map(() =>
-    createClusterRedisClient(
-      config.REDIS_URL,
-      config.REDIS_PASSWORD,
-      config.REDIS_PORT
-    )
-  ),
-  O.getOrElse(() =>
-    createSimpleRedisClient(
-      config.REDIS_URL,
-      config.REDIS_PASSWORD,
-      config.REDIS_PORT,
-      config.REDIS_TLS_ENABLED
+export const redisClientFactory = (): Promise<RedisClient> =>
+  pipe(
+    config.isProduction,
+    O.fromPredicate<boolean>(identity),
```

Comment:
```suggestion
    O.fromPredicate(identity),
```
on my side the type is correctly inferred and it builds.

Reply:
fixed in https://github.com/pagopa/io-functions-cgn/pull/102/commits/1e5923a1d4661210ca8ea4767ed4016fd7ed2481